### PR TITLE
Clean up commit's typedef

### PR DESCRIPTION
**Pull request by hi-rustin**
The parent commit use wrong struct.
You can refer to this API.

https://developer.github.com/v3/repos/commits/

### CODE REVIEW

**Comment by petr-muller** [comment]
Why is this needed? Please start accompanying your PRs with at least minimal description of the change and why is it needed. The subject of this PR says "Fix the type" but the change actually creates a new type. What problem is fixed by this change?

**Comment by hi-rustin** [comment]
> Why is this needed? Please start accompanying your PRs with at least minimal description of the change and why is it needed. The subject of this PR says "Fix the type" but the change actually creates a new type. What problem is fixed by this change?
Thanks for your reply.
Because I want use this parent commit SHA.  The parent commit struct  is wrong, so I want fix it.

**Comment by hi-rustin** [comment]
@petr-muller The PR subject and body updated.

**Comment by hi-rustin** [comment]
/cc @fejta  @alvaroaleman 

Could you please take a look? Thanks!

**Comment by hi-rustin** [comment]
Actually the docs miss the `html_url`.

The actual commit structure is:
```json
  {
    "sha": "3ea0f6572adc49cbaf8df7644bcff2e7f52514e7",
    "node_id": "MDY6Q29tbWl0NTMzMTE3MTY6M2VhMGY2NTcyYWRjNDljYmFmOGRmNzY0NGJjZmYyZTdmNTI1MTRlNw==",
    "commit": {
      "author": {
        "name": "Song Gao",
        "email": "disxiaofei@163.com",
        "date": "2020-10-27T03:13:18Z"
      },
      "committer": {
        "name": "GitHub",
        "email": "noreply@github.com",
        "date": "2020-10-27T03:13:18Z"
      },
      "message": "schedule: simplify the scatter api code (#3078)",
      "tree": {
        "sha": "8153d5566efd46a5e24dc0fee80dda9d1bded17b",
        "url": "https://api.github.com/repos/tikv/pd/git/trees/8153d5566efd46a5e24dc0fee80dda9d1bded17b"
      },
      "url": "https://api.github.com/repos/tikv/pd/git/commits/3ea0f6572adc49cbaf8df7644bcff2e7f52514e7",
      "comment_count": 0,
      "verification": {
        "verified": true,
        "reason": "valid",
        "signature": "-----BEGIN PGP SIGNATURE-----\n\nwsBcBAABCAAQBQJfl5BOCRBK7hj4Ov3rIwAAdHIIAFFYfra1wX7VTOlKQ6wKquI2\nhQlpd5SAW39X2kismkpogOQKq9lV2Q7xsIivLXbyEDB+p2LKPaT2os/Cq860yqJ2\nSUzAPXLpkCTvhtRWwxQGcd8ReYruWKrdWmsEDThmeSzDBFqjKcTtObqPZsRqPfDI\ncwHVshMy4Qsj5Su2FarK+3j3sd9TueqPfquP70IgJAPo9HhAt3PjzAxO7zh7FfKM\n3EBoNxBmSQuFCqtOQLq2z/cWbPaEabBOZKkwOCNTONyfEd38XGPrMD/qaBp/hUPg\nwi9wBW6PzTnwtrcKpZtCeZXfK6ovj6s9PurCzomO71UOTHDKDRUqFSfBxppw7PQ=\n=vyK4\n-----END PGP SIGNATURE-----\n",
        "payload": "tree 8153d5566efd46a5e24dc0fee80dda9d1bded17b\nparent a316b60528b64a1f84a71f87bfdecfc40f81cde2\nauthor Song Gao <disxiaofei@163.com> 1603768398 +0800\ncommitter GitHub <noreply@github.com> 1603768398 +0800\n\nschedule: simplify the scatter api code (#3078)\n\n"
      }
    },
    "url": "https://api.github.com/repos/tikv/pd/commits/3ea0f6572adc49cbaf8df7644bcff2e7f52514e7",
    "html_url": "https://github.com/tikv/pd/commit/3ea0f6572adc49cbaf8df7644bcff2e7f52514e7",
    "comments_url": "https://api.github.com/repos/tikv/pd/commits/3ea0f6572adc49cbaf8df7644bcff2e7f52514e7/comments",
    "author": {
      "login": "Yisaer",
      "id": 13427348,
      "node_id": "MDQ6VXNlcjEzNDI3MzQ4",
      "avatar_url": "https://avatars1.githubusercontent.com/u/13427348?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/Yisaer",
      "html_url": "https://github.com/Yisaer",
      "followers_url": "https://api.github.com/users/Yisaer/followers",
      "following_url": "https://api.github.com/users/Yisaer/following{/other_user}",
      "gists_url": "https://api.github.com/users/Yisaer/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/Yisaer/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/Yisaer/subscriptions",
      "organizations_url": "https://api.github.com/users/Yisaer/orgs",
      "repos_url": "https://api.github.com/users/Yisaer/repos",
      "events_url": "https://api.github.com/users/Yisaer/events{/privacy}",
      "received_events_url": "https://api.github.com/users/Yisaer/received_events",
      "type": "User",
      "site_admin": false
    },
    "committer": {
      "login": "web-flow",
      "id": 19864447,
      "node_id": "MDQ6VXNlcjE5ODY0NDQ3",
      "avatar_url": "https://avatars3.githubusercontent.com/u/19864447?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/web-flow",
      "html_url": "https://github.com/web-flow",
      "followers_url": "https://api.github.com/users/web-flow/followers",
      "following_url": "https://api.github.com/users/web-flow/following{/other_user}",
      "gists_url": "https://api.github.com/users/web-flow/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/web-flow/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/web-flow/subscriptions",
      "organizations_url": "https://api.github.com/users/web-flow/orgs",
      "repos_url": "https://api.github.com/users/web-flow/repos",
      "events_url": "https://api.github.com/users/web-flow/events{/privacy}",
      "received_events_url": "https://api.github.com/users/web-flow/received_events",
      "type": "User",
      "site_admin": false
    },
    "parents": [
      {
        "sha": "a316b60528b64a1f84a71f87bfdecfc40f81cde2",
        "url": "https://api.github.com/repos/tikv/pd/commits/a316b60528b64a1f84a71f87bfdecfc40f81cde2",
        "html_url": "https://github.com/tikv/pd/commit/a316b60528b64a1f84a71f87bfdecfc40f81cde2"
      }
    ]
  }
```

**Review by petr-muller — APPROVED** [review]
The lack of documentation of that structure in GH API does not help, but this looks correct.

**Comment by hi-rustin** [comment]
@fejta @alvaroaleman Could you please take a look? Thanks!

File: prow/github/types.go [hunk 1]
```diff
@@ -1090,14 +1090,21 @@ type Milestone struct {
...
+	Author      User           `json:"author"`
+	Committer   User           `json:"committer"`
+	Parents     []ParentCommit `json:"parents,omitempty"`
+	HTMLURL     string         `json:"html_url"`
+	URL         string         `json:"url"`
+	CommentsURL string         `json:"comments_url"`
+}
+
+// ParentCommit represents the parent of commit.
+type ParentCommit struct {
```
Comment (alvaroaleman) [inline prow/github/types.go:1104]: Can you copy the full typedef from https://github.com/google/go-github/blob/b47a3f75fc03ed477be4f04813180deb31407570/github/git_commits.go#L24 and make all scalar types literals and not pointers?

Comment (hi-rustin) [reply prow/github/types.go:1104]: I think this is unreasonable, because it has too many fields we don't need.

Comment (hi-rustin) [reply prow/github/types.go:1104]: And I think the structure of your link is not suitable for repo commit. You can check my previous [comment](https://github.com/kubernetes/test-infra/pull/19834#issuecomment-722966106).

Comment (alvaroaleman) [reply prow/github/types.go:1104]: The [github docs](https://developer.github.com/v3/repos/commits/) mention all of the fields in that type, and its just as (un-)reasonable as your initial change, we don't use any of these fields

Comment (hi-rustin) [reply prow/github/types.go]: Now the `RepositoryCommit` same as https://github.com/google/go-github/blob/b47a3f75fc03ed477be4f04813180deb31407570/github/repos_commits.go#L18 and `Commit` same as https://github.com/google/go-github/blob/b47a3f75fc03ed477be4f04813180deb31407570/github/git_commits.go#L24.

File: prow/github/types.go [hunk 2]
```diff
@@ -813,10 +813,18 @@ func (pe PushEvent) Branch() string {
 	return ref
 }
 
-// Commit represents general info about a commit.
-type Commit struct {
+// HeadCommit represents a git commit in a GitHub PushEvent.
+type HeadCommit struct {
```
Comment (alvaroaleman) [inline prow/github/types.go:817]: Can we please keep the name of the type?

File: prow/github/types.go [hunk 3]
```diff
@@ -1090,14 +1101,70 @@ type Milestone struct {
...
+	Commit      Commit   `json:"commit,omitempty"`
+	Author      User     `json:"author,omitempty"`
+	Committer   User     `json:"committer,omitempty"`
+	Parents     []Commit `json:"parents,omitempty"`
+	HTMLURL     string   `json:"html_url,omitempty"`
+	URL         string   `json:"url,omitempty"`
+	CommentsURL string   `json:"comments_url,omitempty"`
+
+	// Details about how many changes were made in this commit. Only filled in during GetCommit!
+	Stats CommitStats `json:"stats,omitempty"`
```
Comment (alvaroaleman) [inline prow/github/types.go:1115]: Please make this a pointer, otherwise it will always be present

File: prow/github/types.go [hunk 4]
```diff
@@ -1090,14 +1101,70 @@ type Milestone struct {
...
+	SHA          string                `json:"sha,omitempty"`
+	Author       User                  `json:"author,omitempty"`
+	Committer    User                  `json:"committer,omitempty"`
+	Message      string                `json:"message,omitempty"`
+	Tree         Tree                  `json:"tree,omitempty"`
+	Parents      []Commit              `json:"parents,omitempty"`
+	Stats        CommitStats           `json:"stats,omitempty"`
+	HTMLURL      string                `json:"html_url,omitempty"`
+	URL          string                `json:"url,omitempty"`
+	Verification SignatureVerification `json:"verification,omitempty"`
```
Comment (alvaroaleman) [inline prow/github/types.go:1153]: Please make this a pointer, because it might be empty

File: prow/github/types.go (hunk 2 above)
Comment (hi-rustin) [reply prow/github/types.go:817]: We have a new type name also called Commit. The current name is the same as go-github.

File: prow/github/types.go (hunk 4 above)
Comment (hi-rustin) [reply prow/github/types.go]: Fixed.

File: prow/github/types.go (hunk 3 above)
Comment (hi-rustin) [reply prow/github/types.go:1115]: Fixed.

Comment (hi-rustin) [reply prow/github/types.go]: Fixed.

File: prow/github/types.go (hunk 2 above)
Comment (hi-rustin) [reply prow/github/types.go]: Renamed. PTAL~ @alvaroaleman

Comment (hi-rustin) [reply prow/github/types.go]: Now the `RepositoryCommit` same as https://github.com/google/go-github/blob/b47a3f75fc03ed477be4f04813180deb31407570/github/repos_commits.go#L18 and `GitCommit` same as https://github.com/google/go-github/blob/b47a3f75fc03ed477be4f04813180deb31407570/github/git_commits.go#L24.

Comment (hi-rustin) [reply prow/github/types.go]: I don't know should I keep the old comment for `Commit`, but I think the current comment is more accurate.

Comment (hi-rustin) [reply prow/github/types.go]: @alvaroaleman ping~

**Comment by alvaroaleman** [comment]
@hi-rustin , sorry for the delay but we need to be extra careful with a change like this and right now I am a bit busy

**Comment by hi-rustin** [comment]
> @hi-rustin , sorry for the delay but we need to be extra careful with a change like this and right now I am a bit busy

Got it. Thanks for your reply and review. Because I have some work block by it, so I really hope it can be landed. But we should be more careful and hope that it doesn’t disturb you.

**Comment by hi-rustin** [comment]
I think the `SingleCommit` same as `RepositoryCommit`, so I dropped it.

File: prow/github/types.go [hunk 5]
```diff
@@ -1090,20 +1094,70 @@ type Milestone struct {
...
 type RepositoryCommit struct {
-	SHA         string    `json:"sha"`
-	Commit      GitCommit `json:"commit"`
-	Author      User      `json:"author"`
-	Committer   User      `json:"committer"`
-	Parents     []Commit  `json:"parents,omitempty"`
-	HTMLURL     string    `json:"html_url"`
-	URL         string    `json:"url"`
-	CommentsURL string    `json:"comments_url"`
+	NodeID      string      `json:"node_id,omitempty"`
```
Comment (alvaroaleman) [inline prow/github/types.go:1097]: @hi-rustin please scope this PR back down what you initially had, extending the `RepositoryCommit` structs and its elements. If you want to do further changes to the types here, you can create follow-ups but then:
* Provide a reference to the github api docs for the relevant struct and add it as a godoc comment, if applicable
* Provide a sample of where this struct is used, so it becomes clear that its actually used for the referenced api endpoint

Changes in the gh types like this have a huge blast radius, so if you want a chance at getting them merged you need to:
* Keep them well-scoped
* Convince me that this change is safe (the ever-expanding scope of this PR combined with the lack of references is the main reason its sitting so long, going through our codebase and the github api docs in order to determine if a change I personally don't need is safe is very low on my priority list)

Comment (hi-rustin) [reply prow/github/types.go:1096]: Reverted. How about now?

Comment (hi-rustin) [reply prow/github/types.go:1096]: Now all structs copy from go-gihub.

File: prow/github/types.go [hunk 6]
```diff
@@ -1090,20 +1093,70 @@ type Milestone struct {
...
+	ContentsURL      string `json:"contents_url,omitempty"`
+	PreviousFilename string `json:"previous_filename,omitempty"`
 }
 
 // GitCommit represents a GitHub commit.
 type GitCommit struct {
-	SHA     string `json:"sha,omitempty"`
-	Message string `json:"message,omitempty"`
+	SHA          string                 `json:"sha,omitempty"`
+	Author       User                   `json:"author,omitempty"`
```
Comment (alvaroaleman) [inline prow/github/types.go:1137]: It doesn't look like `User` is the right struct here: https://github.com/google/go-github/blob/1e6d9a5f3da386874065352750ee38a12f2efb9f/github/git_commits.go#L30

File: prow/github/types.go [hunk 7]
```diff
@@ -1090,20 +1093,70 @@ type Milestone struct {
...
+	Committer    User                   `json:"committer,omitempty"`
```
Comment (alvaroaleman) [inline prow/github/types.go:1138]: same as above

File: prow/github/types.go [hunk 8]
```diff
@@ -1090,20 +1093,70 @@ type Milestone struct {
...
+	Stats        *CommitStats           `json:"stats,omitempty"`
+	HTMLURL      string                 `json:"html_url,omitempty"`
+	URL          string                 `json:"url,omitempty"`
+	Verification *SignatureVerification `json:"verification,omitempty"`
+	NodeID       string                 `json:"node_id,omitempty"`
+
+	// CommentCount is the number of GitHub comments on the commit. This
+	// is only populated for requests that fetch GitHub data like
+	// Pulls.ListCommits, Repositories.ListCommits, etc.
+	CommentCount int `json:"comment_count,omitempty"`
```
Comment (alvaroaleman) [inline prow/github/types.go:1151]: Please make this one a `*int` so clients can differentiate if it was set or not

Comment (hi-rustin) [reply prow/github/types.go]: Added.

File: prow/github/types.go (hunk 6 above)
Comment (hi-rustin) [reply prow/github/types.go]: Good catch! Fixed.

File: prow/github/types.go [hunk 9]
```diff
@@ -1090,20 +1093,81 @@ type Milestone struct {
...
+
+// CommitAuthor represents the author or committer of a commit. The commit
+// author may not correspond to a GitHub User.
+type CommitAuthor struct {
+	Date  time.Time `json:"date,omitempty"`
+	Name  string    `json:"name,omitempty"`
+	Email string    `json:"email,omitempty"`
+
+	// The following fields are only populated by Webhook events.
+	Login string `json:"username,omitempty"`
```
Comment (alvaroaleman) [inline prow/github/types.go:1162]: Please make this a pointer as its optional

Comment (hi-rustin) [reply prow/github/types.go]: Sorry for that, I'm  too careless.

**Review by alvaroaleman — APPROVED** [review]
thanks!

**Comment by hi-rustin** [comment]
> thanks!

Thank you again for your patience and review. ❤️

**Pull request merged**